### PR TITLE
feat: Auto find a suitable neovim executable

### LINKED ISSUE #195: Feature request: New install experience
I would like to improve on the first-install experience.

1. Have vscode-neovim.neovimPath be 'nvim' and search $PATH for it by default.
2. If nvim is found but have the wrong version. ERROR OUT and show an error on start.

This plugin requires me to always open up the config to use the tool when I install vscode with the neovim ext. on a new computer. Which I do.. often! And equally often I will screw up on the configuration and wonder why all keyboard input is being dropped. After several restarts and WTFs I realize that "oh its called nvim not neovim" or "oh neovim was old on this system I need to install it out of tree" or something else.

### PR DESCRIPTION
**Pull request by xiyaowong**
Close #195

### CODE REVIEW

File: src/main_controller.ts [hunk 1]
```diff
@@ -184,8 +186,21 @@ export class MainController implements vscode.Disposable {
             }
         }
 
+        let neovimPath = config.neovimPath;
+        // Only try to find nvim if the path is the default one
+        // And if we are not using WSL
```
Comment (justinmk) [inline src/main_controller.ts:191]: could findNvim() be enhanced to work with WSL? PR welcome at https://github.com/neovim/node-client/

File: src/main_controller.ts [hunk 2]
```diff
@@ -184,8 +186,21 @@ export class MainController implements vscode.Disposable {
...
+        if (neovimPath === "nvim" && !config.useWsl) {
+            const nvimResult = findNvim({ minVersion: NVIM_MIN_VERSION });
+            logger.debug("Find nvim result: ", nvimResult);
```
Comment (justinmk) [inline src/main_controller.ts:194]: should be info()

File: src/main_controller.ts [hunk 3]
```diff
@@ -104,6 +105,7 @@ export class MainController implements vscode.Disposable {
         this.client.on("request", this.onNeovimRequest);
         this.setClientInfo();
         await this.client.setVar("vscode_channel", await this.client.channelId);
+        await this.client.setVar("vscode_nvim_min_version", NVIM_MIN_VERSION);
```
Comment (justinmk) [inline src/main_controller.ts:108]: instead of adding more global vars can we change `g:vscode` to be a dict, and store things in its keys

**Review by justinmk — APPROVED** [review]
I have a local branch that does this, but this is fine for now.

`findNvim()` might change to async, that's one reason why I didn't do this yet

File: src/main_controller.ts (hunk 1 above)
Comment (xiyaowong) [reply src/main_controller.ts:191]: I personally think it's not very meaningful. The use case for nvim in WSL is limited, and WSL essentially functions like a virtual machine.

File: src/main_controller.ts (hunk 3 above)
Comment (xiyaowong) [reply src/main_controller.ts:108]: But it’s not convenient to update some data, right? Sometimes variables are not defined at the same time

**Comment by xiyaowong** [comment]
The time taken by findNvim isn't too long, which is acceptable. However, I hope findNvim can provide an option to stop searching when it finds a valid nvim. 🤔

File: src/main_controller.ts (hunk 3 above)
Comment (justinmk) [reply src/main_controller.ts:108]: untested, but just calling `vim.g.vscode = vim.g.vscode or {}` first is a good enough pattern (eventually Nvim core should make this easier):

```typescript
await this.client.lua("vim.g.vscode = vim.g.vscode or {}");
await this.client.lua("vim.g.vscode.channel = ...", await this.client.channelId);
await this.client.lua("vim.g.vscode.nvim_min_version", = ...", NVIM_MIN_VERSION);
```

Comment (xiyaowong) [reply src/main_controller.ts:108]: In this case, there are quite a few places that need to be modified. If refactoring is required, a separate PR would be more appropriate.

Comment (xiyaowong) [reply src/main_controller.ts:108]: > Untested:

I just tested it and it doesn't work properly

Comment (justinmk) [reply src/main_controller.ts:108]: > I just tested it and it doesn't work as expected.

what exactly does that mean?

Comment (xiyaowong) [reply src/main_controller.ts:108]: image

Comment (justinmk) [reply src/main_controller.ts:108]: Oh that's because the container is marshalled. So it will need to be vimscript intsead of lua.

**Pull request merged**